File: algorithms/implementation/grading/solution-fp.php
<?php

namespace numbers {

    function next_int($start, $divisible_by) {
        $num = $start;
        while ($num % $divisible_by !== 0) {
            $num += 1;
        }
        return $num;
    }

}

namespace grades {

    const FAILING_GRADE = 40;

    function round_grade($grade) {
        $possible = \numbers\next_int($grade, 5);
        $rounded = $grade;
        if (($possible - $grade) < 3) {
            $rounded = $possible;
        }
        return $rounded >= FAILING_GRADE ? $rounded : $grade;
    }

    function round_grades($grades) {
        return array_map('\grades\round_grade', $grades);
    }

}

namespace grades\io {

    function parse_grades($in_handle) {
        fscanf($in_handle, "%d", $student_count);
        assert($student_count >= 1 && $student_count <= 60);

        foreach (range(1, $student_count) as $i) {
            fscanf($in_handle, "%d", $grade);
            assert($grade >= 0 && $grade <= 100);
            $grades[] = $grade;
        }

        return $grades;
    }

    function print_grades($out_handle, $grades) {
        foreach ($grades as $grade) {
            fprintf($out_handle, "%d\n", $grade);
        }
    }

}

namespace main {

    function main() {
        $in_handle = fopen('php://stdin', 'r');
        $out_handle = fopen('php://stdout', 'w');

        try {
            $grades = \grades\io\parse_grades($in_handle);
            $rounded = \grades\round_grades($grades);
            \grades\io\print_grades($out_handle, $rounded);
        }
        finally {
            fclose($out_handle);
            fclose($in_handle);
        }
    }

    main();

}

File: algorithms/implementation/grading/solution-fp.py
#!/bin/python3

import sys
from typing import List


Grade = int


FAILING_GRADE = 40


def parse_grade(raw: str) -> Grade:
    grade = Grade(raw)
    assert 0 <= grade <= 100
    return grade


def parse_grades(input_file) -> List[Grade]:
    n = int(input_file.readline())
    assert 1 <= n <= 60
    return [parse_grade(input_file.readline()) for _ in range(n)]


def print_grades(output_file, grades: List[Grade]) -> None:
    for grade in grades:
        output_file.write(str(grade))
        output_file.write('\n')


def next_number(start: int, divisible_by: int) -> int:
    num = start
    while num % divisible_by != 0:
        num += 1
    return num


def round_grade(grade: Grade) -> Grade:
    possible = next_number(start=grade, divisible_by=5)
    rounded = grade
    if (possible - grade) < 3:
        rounded = possible
    return rounded if rounded >= FAILING_GRADE else grade


def round_grades(grades: List[Grade]) -> List[Grade]:
    return [round_grade(grade) for grade in grades]


if __name__ == '__main__':
    grades = parse_grades(sys.stdin)
    rounded_grades = round_grades(grades)
    print_grades(sys.stdout, rounded_grades)
